accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+
+We will need to keep track of some information for every wasm opcode in our
+program: what program counter the opcode is, the size of its operand stack, and
+which entry in the operand stack the opcode is in terms of machine state. All of
+this information can be serialized and passed into the utility function present
+in the `<check for breakpoint>` pseudo opcodes above.
+
+### Watchpoints
+
+Watchpoints for locals, globals, and table elements can be converted to
+breakpoints by statically checking for local.set, global.set, and table.set
+instructions. Memory watchpoints— pausing execution when a program writes to a
+specific memory location— are a little bit trickier to accomplish. However, like
+the proposed implementation for single-stepping mentioned in the [previous
+section](#breakpoints-and-single-stepping), we can implement watch-points by
+instrumenting all store operations with a similar call out to a utility
+function. The call sites (store instructions) would have enough local
+information available to determine if a watchpoint would trigger, and could
+provide us with another implementation path that would avoid modifying the
+signal handler.
+
+### Step into / out of
+
+Stepping into a function means pausing execution on the first instruction
+_inside_ a function that is about to be called. We can accomplish this by
+setting a temporary breakpoint on that first instruction, but first we need to
+figure out the address of the function we are about to step into. For direct
+function calls, we can set a breakpoint at the start of the static callee. For
+indirect calls, as long as we pause immediately before the indirect branch, we
+can see which func index we are about to call and set a breakpoint accordingly.
+
+Alternatively, as we are already inserting various bookkeeping function calls
+elsewhere in the function body during code generation, we could track one bit of
+global information that indicates whether or not a function should break
+immediately on entry, checking and potentially clearing that bit at every
+function entry. This would allow us to step into functions by setting that bit
+while in single-stepping mode before processing a call instruction, and then
+clearing that bit once the next function has been entered. This implementation
+adds some overhead by requiring that we are already stepping before the call
+instruction is processed, but does simplify implementation details by not
+requiring special handling of indirect calls.
+
+Stepping out of a function is a bit easier to implement, as it will require us
+to stop execution at any instruction that could cause the function to return
+(branches with a sufficiently large stack depth, or a return instruction), and
+then single-step out. For tail calls we have multiple options: we could treat
+stepping out as breaking when the next function is entered, or we could break
+once we finally return from the tail callee.
```

Comment:
Thinking out loud here; this is not a fully-fleshed out thought.

If this `debug_trap()` function is implemented like this pseudo-code:

```python
def debug_trap(reason):
    mask = global_debug_trap_mask & reason
    if mask & (TRAP_ON_FUNCTION_RETURN | TRAP_ON_INSTRUCTION | ...):
        platform_debug_trap()
```

Then this makes it possible to select the kind of single-stepping we're doing by setting/resetting bits in `global_debug_trap_mask` or implementing anything specific for stepping into or out of functions. This incurs in a bit more overhead than just checking a boolean flag, but allows a bit more flexibility that can help the implementation.

Can even have things like `debug_trap(TRAP_ON_END_OF_LOOP)` emitted after every loop to skip long, boring loops while single-stepping code.